### DISCUSSION TITLE
How to write a plugin?

### DISCUSSION THREAD

**Discussion by jantimon** [swc-project/swc#2268] [Q&A] [answered]
Is there any guide to write a plugin?

All I found was a guide on howto build deprecated plugins: https://swc.rs/docs/usage-plugin

**Comment by aelbore** [comment]
@jantimon hope this helps https://github.com/aelbore/lit-swc-transpiler
i dont have docs for it, let me know if you have any questions :)

**Reply by kdy1** [reply]
No. Rust-based plugin api is still WIP, mainly because I'm busy working on proprietary software.

**Reply by jantimon** [reply]
Thanks for your quick response :) 

is there any timeline for the rust api?

I don't want to invest time writing ts or js code if it won't work anymore for 2.0 soon

**Reply by kdy1** [reply]
I don't have any timeline because of time.
Open-source is not rewarding, so I want to focus on some closed-source projects.

**Reply by jantimon** [reply]
I am sorry to hear and can totally understand what you say.. 

In that case I'll use the JS api to play around with swc & swc plugins for now and will rewrite it once there is a 2.0 release

Take care :)

**Reply by kdy1** [reply]
For clarification. (Sorry for my bad English skill)

I'll keep working on open source. The difference is that previously I used almost all of my free time for swc, but I'm going to use some time for closed-source projects.

**Comment by kdy1** [comment]
I've been thinking about some issues like AST compat problem, and I decided to just implement plugin system under experimental flags and see what happens.

**Reply by kdy1** [reply]
So the estimate is 1 month at max.

**Reply by jantimon** [reply]
Wow that sounds amazing - one of my really small OSS projects ships with a babel plugin.. as soon as you reach any state where you need real world feedback let me know and I'll try to port it :)

**Comment by vjpr** [comment]
Is the plugin API going to be much different to the current Rust internal visitor implementation (e.g. how it is used here: https://github.com/vercel/next.js/blob/canary/packages/next/build/swc/src/styled_jsx/transform_css.rs)?

And do you intend to add convenience helper methods like Babel had?

**Reply by kdy1** [reply]
The `swc_plugin` crate will re-export required types.
Helper methods will be provided via direct dependency to `swc_ecma_utils`  from plugin crates, to reduce breaking change of `swc_plugin`.

**Comment by kwonoj** [comment, marked as answer]
Please refer https://github.com/swc-project/swc/discussions/3540 and try new plugin interfaces. We expect this to be an official recommended approach once stablized.
